### PR TITLE
Feature/add phpstan

### PR DESCRIPTION
**Pull request by DeepDiver1975**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,23 +1,38 @@
 language: php
 sudo: required
 php:
-  - 7.0
-  - 7.1
-  - 7.2
-  - 7.3
+- 7.0
```
Comment (staabm) [inline .travis.yml]: Re-indent

File: .travis.yml [hunk 2]
```diff
@@ -1,23 +1,38 @@
...
+  - RUN_PHPSTAN="FALSE"
 
 matrix:
+  include:
+  - name: 'PHPStan'
+    php: 7.2
+    env: RUN_PHPSTAN="TRUE"
   fast_finish: true
+  allow_failures:
+  - php: 5.5
```
Comment (staabm) [inline .travis.yml]: Does this make sense? We no longer have a 5.5 build?

File: lib/ContextStackTrait.php [hunk 3]
```diff
@@ -45,7 +45,7 @@
      * common use-case for parsing XML documents, it's added here as a
      * convenience.
      *
-     * @var string
+     * @var string | null
```
Comment (staabm) [inline lib/ContextStackTrait.php]: are spaces allowed arround the pipe?

File: lib/Deserializer/functions.php [hunk 4]
```diff
@@ -298,6 +298,8 @@ function mixedContent(Reader $reader): array
  * You can use, e.g., a named constructor (factory method) to create an object using
  * this function.
  *
+ * @param callable $func
```
Comment (staabm) [inline lib/Deserializer/functions.php]: Do we really need this, since it does not provide additional value?

File: lib/Deserializer/functions.php [hunk 5]
```diff
@@ -330,5 +332,5 @@ function functionCaller(Reader $reader, callable $func, string $namespace)
     } while (Reader::END_ELEMENT !== $reader->nodeType);
     $reader->read();
 
-    return $func(...array_values($funcArgs));
+    return call_user_func_array($func, array_values($funcArgs));
```
Comment (staabm) [inline lib/Deserializer/functions.php]: why this change?

File: lib/Deserializer/functions.php [hunk 6]
```diff
@@ -309,8 +311,8 @@ function functionCaller(Reader $reader, callable $func, string $namespace)
     }
 
     $funcArgs = [];
-    $func = is_string($func) && false !== strpos($func, '::') ? explode('::', $func) : $func;
-    $ref = is_array($func) ? new \ReflectionMethod($func[0], $func[1]) : new \ReflectionFunction($func);
+    $func0 = is_string($func) && false !== strpos($func, '::') ? explode('::', $func) : $func;
```
Comment (staabm) [inline lib/Deserializer/functions.php]: why this rename?

Comment (DeepDiver1975) [reply lib/Deserializer/functions.php]: to make phpstan happy

File: lib/Deserializer/functions.php (hunk 5 above)
Comment (DeepDiver1975) [reply lib/Deserializer/functions.php]: to make phpstan happy

File: lib/Deserializer/functions.php (hunk 4 above)
Comment (DeepDiver1975) [reply lib/Deserializer/functions.php]: to make phpstan happy

File: lib/ContextStackTrait.php (hunk 3 above)
Comment (DeepDiver1975) [reply lib/ContextStackTrait.php]: yes

File: lib/Deserializer/functions.php (hunk 4 above)
Comment (DeepDiver1975) [reply lib/Deserializer/functions.php]: FYI: I moved the level down to 6 and removed all these changes - I was anyhow not that convinced ....

Comment (staabm) [reply lib/Deserializer/functions.php]: Thx. Agree on this.

**Comment by DeepDiver1975** [comment]
I'll squash the commits together and merge then ....
